### PR TITLE
feature/5008-Finalize-PATCH-Action-Logic

### PR DESCRIPTION
**Pull request by bcdevlucas**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namex license (Apache 2.0).

### CODE REVIEW

File: api/namex/models/request.py [hunk 1]
```diff
@@ -111,6 +111,19 @@ class Request(db.Model):
     REQUEST_FURNISHED = 'Y'
 
     # properties
+    @property
+    def has_consumed_name(self):
```
Comment (LJTrent) [inline api/namex/models/request.py:115]: hmm. wondering why this is not part of the names model instead?

File: api/namex/models/request.py [hunk 2]
```diff
@@ -111,6 +111,19 @@ class Request(db.Model):
...
+        names = self.names.all()
+
+        def is_not_consumed(name):
```
Comment (LJTrent) [inline api/namex/models/request.py:118]: same thing, should this be in the names model?

File: api/namex/resources/name_requests/api_models.py [hunk 3]
```diff
@@ -33,7 +33,8 @@
     'designation': fields.String('Name designation based on entity type'),
     'conflict1_num': fields.String('The corp_num of the matching name'),
     'conflict1': fields.String('The matching corp name'),
-    'consent_words': fields.Nested(consent_model)
+    'consent_words': fields.Nested(consent_model),
+    'corpNum': fields.String('Corp Num')
```
Comment (LJTrent) [inline api/namex/resources/name_requests/api_models.py:37]: what about  consumptionDate

File: api/namex/resources/name_requests/name_request.py [hunk 4]
```diff
@@ -267,20 +267,37 @@ def patch(self, nr_id, nr_action):
             nr_svc.nr_num = nr_model.nrNum
             nr_svc.nr_id = nr_model.id
 
-            valid_states = State.VALID_STATES
+            # Only allow editing if the request is in certain valid states
+            request_editable_states = [
+                State.DRAFT,
+                State.RESERVED,
```
Comment (LJTrent) [inline api/namex/resources/name_requests/name_request.py:273]: Why is reserved and cond-reserve editable.  We dont want to edit those from the patch?

File: api/namex/models/request.py (hunk 1 above)
Comment (bcdevlucas) [reply api/namex/models/request.py:115]: Convenience eg.
`
def display_cancel_action(nr_model=None):
    try:
        if (nr_model and nr_model.stateCd == State.CANCELLED) or (nr_model.stateCd not in State.CANCELLABLE_STATES):
            return False

        if nr_model and (nr_model.is_expired or nr_model.has_consumed_name):
            return False

        return True
    except Exception as err:
        raise NameRequestActionError(err)
`

File: api/namex/models/request.py (hunk 2 above)
Comment (bcdevlucas) [reply api/namex/models/request.py:118]: Same as my comment above.

File: api/namex/resources/name_requests/api_models.py (hunk 3 above)
Comment (bcdevlucas) [reply api/namex/resources/name_requests/api_models.py:37]: Not sure about consumptionDate... I didn't see that in the ticket.

File: api/namex/resources/name_requests/name_request.py (hunk 4 above)
Comment (bcdevlucas) [reply api/namex/resources/name_requests/name_request.py:273]: Maybe editable isn't the best word... I'm just checking to see what states should actually be 'handled' within the PATCH method.

File: api/namex/resources/name_requests/api_models.py (hunk 3 above)
Comment (LJTrent) [reply api/namex/resources/name_requests/api_models.py:37]: consumptionDate and corpNum define this as a consumed name.

**Pull request merged**